core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,397 @@
+
+  def normalizeAnnotation(a: Annotation[Declaration], env: Env, p: Package.Inferred):
+    NormState[TypedExpr[(Declaration, NormalExpressionTag)]] =
+      normalizeExpr(a.term, env, p).map { term =>
+        val neTag = term.tag._2
+        val tag = (a.tag, neTag)
+        a.copy(term=term, tag=tag)
+      }
+
+  def normalizeGeneric(g: Generic[Declaration], env: Env, p: Package.Inferred): 
+    NormState[TypedExpr[(Declaration, NormalExpressionTag)]] =
+      normalizeExpr(g.in, env, p).map { in =>
+        val neTag = in.tag._2
+        val tag = (g.tag, neTag)
+        g.copy(in=in, tag=tag)
+      }
+
+  def normalizeVar(v: Var[Declaration], env: Env, p: Package.Inferred):
+    NormState[TypedExpr[(Declaration, NormalExpressionTag)]] =
+      env._1.get(v.name) match {
+        case None => norm((p, Left((v.name, v.tag)), env)).map {ne =>
+          val neTag = getTag(ne)._2
+          v.copy(tag=(v.tag, neTag))
+        }
+        case Some(neTag) =>
+          State.pure(v.copy(tag=(v.tag, neTag)))
+      }
+
+  def normalizeAnnotatedLambda(al: AnnotatedLambda[Declaration], env: Env, p: Package.Inferred):
+    NormState[AnnotatedLambda[(Declaration, NormalExpressionTag)]] = {
+      val lambdaVars = Some(al.arg) :: env._2
+      val nextEnv: Env = (env._1 ++ lambdaVars.zipWithIndex
+        .collect { case (Some(n), i) => (n, i) }
+        .toMap
+        .mapValues(idx => NormalExpressionTag(NormalExpression.LambdaVar(idx), Set[NormalExpression]())),
+        lambdaVars)
+      for {
+        eExpr <- normalizeExpr(al.expr, nextEnv, p)
+        ne = NormalExpression.Lambda(eExpr.tag._2.ne)
+        children = combineWithChildren(eExpr.tag._2)
+        neTag = NormalExpressionTag(ne, children)
+      } yield al.copy(expr=eExpr, tag=(al.tag, neTag))
+    }
+
+  def normalizeApp(a: App[Declaration], env: Env, p: Package.Inferred):
+    NormState[App[(Declaration, NormalExpressionTag)]] =
+      for {
+        efn <- normalizeExpr(a.fn, env, p)
+        earg <- normalizeExpr(a.arg, env, p)
+        ne = NormalExpression.App(efn.tag._2.ne, earg.tag._2.ne)
+        children = combineWithChildren(efn.tag._2) ++ combineWithChildren(earg.tag._2)
+        neTag = NormalExpressionTag(ne, children)
+      } yield a.copy(fn=efn, arg=earg, tag=(a.tag, neTag))
+
+  def normalizeLet(l: Let[Declaration], env: Env, p: Package.Inferred):
+    NormState[TypedExpr[(Declaration, NormalExpressionTag)]] = {
+      val (nextEnv, neWrapper) = l.recursive match {
+        case RecursionKind.Recursive => {
+          val lambdaVars = Some(l.arg) :: env._2
+          val nextEnv = (env._1 ++ lambdaVars.zipWithIndex
+            .collect { case (Some(n), i) => (n, i) }
+            .toMap
+            .mapValues(idx => NormalExpressionTag(NormalExpression.LambdaVar(idx), Set[NormalExpression]())),
+            lambdaVars)
+          val neWrapper = {ne: NormalExpression => NormalExpression.Recursion(NormalExpression.Lambda(ne)).asInstanceOf[NormalExpression]}
```

Comment:
can we remove the cast? I think by adding a `: NormalExpression` is all that is needed (ascribing a type, vs casting).

Reply:
Ohhh... I was busy looking at covariant and contravariant type variables. That makes way more sense